provider/datagen/src/transform/cldr/units/mod.rs

```diff
@@ -20,16 +31,103 @@ impl DataProvider<UnitsConstantsV1Marker> for crate::DatagenProvider {
-            constants_map: ZeroMap::from_iter(constants_map),
-        };
+        // This loop will replace all the constants in the value of a constant with their values.
+        loop {
+            let mut cons_with_text: u16 = 0;
+            for (cons_name, (num, den, constant_type)) in constants_with_constants_map.iter() {
+                for i in 0..num.len() {
+                    if !has_letters(num[i]) {
+                        continue;
+                    }
+
+                    if constants_with_constants_map.contains_key(num[i]) {
+                        let (rnum, rden, rconstant_type) =
+                            constants_with_constants_map.get(num[i]).unwrap();
+                        num.remove(i);
+                        // append the elements in rnum to num and rden to den
+                        num.append(&mut rnum.clone());
+                        den.append(&mut rden.clone());
+
+                        if *rconstant_type == ConstantType::Approximate {
+                            *constant_type = ConstantType::Approximate;
+                        }
+                    } else {
+                        cons_with_text += 1;
+                    }
+                }
+
+                for i in 0..den.len() {
+                    if !has_letters(den[i]) {
+                        continue;
+                    }
+
+                    if constants_with_constants_map.contains_key(den[i]) {
+                        let (rnum, rden, constant_type) =
+                            constants_with_constants_map.get(den[i]).unwrap();
+                        den.remove(i);
+                        // append the elements in rden to num and rnum to den
+                        num.append(&mut rden.clone());
+                        den.append(&mut rnum.clone());
+
+                        if *constant_type == ConstantType::Approximate {
+                            *constant_type = ConstantType::Approximate;
+                        }
+                    } else {
+                        cons_with_text += 1;
+                    }
+                }
+            }
+
+            if cons_with_text == 0 {
+                break;
+            }
+        }
+
+        let mut constants_map = BTreeMap::<&str, ConstantValue>::new();
+
+        for (cons_name, (num, den, constant_type)) in constants_with_constants_map.iter() {
+            let value = convert_array_of_strings_to_fraction(num, den)?;
+            let (num, den, sign, cons_type) =
+                convert_fractional_to_constant_value(value, *constant_type)?;
+            constants_map.insert(
+                cons_name,
+                ConstantValue {
+                    numerator: ZeroVec::from_iter(num),
+                    denominator: ZeroVec::from_iter(den),
+                    sign,
+                    constant_type: cons_type,
+                },
+            );
+        }
+
+        let constants_map = ZeroMap::from_iter(constants_map.into_iter());
```

Comment:
@sffc ,

I got the following error:

```
[cargo-make] INFO - Execute Command: "cargo" "test" "-p=icu_datagen" "--no-default-features" "--features=fs_exporter,use_wasm,rayon,experimental_components" "generate_json_and_verify_postcard" "--" "--nocapture"
   Compiling icu_datagen v1.2.5 (/Users/younies/icu/icu4x/provider/datagen)
warning: unused import: `fraction::GenericFraction`
  --> provider/datagen/src/transform/cldr/units/mod.rs:10:5
   |
10 | use fraction::GenericFraction;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: unused import: `ConstantValueULE`
  --> provider/datagen/src/transform/cldr/units/mod.rs:16:34
   |
16 |     ConstantType, ConstantValue, ConstantValueULE, UnitsConstantsV1, UnitsConstantsV1Marker,
   |                                  ^^^^^^^^^^^^^^^^

warning: unused import: `num_bigint::BigUint`
  --> provider/datagen/src/transform/cldr/units/mod.rs:18:5
   |
18 | use num_bigint::BigUint;
   |     ^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `ConstantValue<'_>: ZeroMapKV<'_>` is not satisfied
   --> provider/datagen/src/transform/cldr/units/mod.rs:129:29
    |
129 |         let constants_map = ZeroMap::from_iter(constants_map.into_iter());
    |                             ^^^^^^^^^^^^^^^^^^ the trait `ZeroMapKV<'_>` is not implemented for `ConstantValue<'_>`
    |
    = help: the following other types implement trait `ZeroMapKV<'a>`:
              char
              i8
              i16
              i32
              i64
              i128
              usize
              u8
            and 72 others
    = note: required for `ZeroMap<'_, _, ConstantValue<'_>>` to implement `FromIterator<(&str, ConstantValue<'_>)>`

error[E0277]: the trait bound `ConstantValue<'_>: ZeroMapKV<'_>` is not satisfied
   --> provider/datagen/src/transform/cldr/units/mod.rs:129:29
    |
129 |         let constants_map = ZeroMap::from_iter(constants_map.into_iter());
    |                             ^^^^^^^ the trait `ZeroMapKV<'_>` is not implemented for `ConstantValue<'_>`
    |
    = help: the following other types implement trait `ZeroMapKV<'a>`:
              char
              i8
              i16
              i32
              i64
              i128
              usize
              u8
            and 72 others
note: required by a bound in `ZeroMap`
   --> /Users/younies/icu/icu4x/utils/zerovec/src/map/map.rs:64:8
    |
61  | pub struct ZeroMap<'a, K, V>
    |            ------- required by a bound in this struct
...
64  |     V: ZeroMapKV<'a> + ?Sized,
    |        ^^^^^^^^^^^^^ required by this bound in `ZeroMap`

error[E0277]: the trait bound `ConstantValue<'_>: ZeroMapKV<'_>` is not satisfied
   --> provider/datagen/src/transform/cldr/units/mod.rs:129:29
    |
129 |         let constants_map = ZeroMap::from_iter(constants_map.into_iter());
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `ZeroMapKV<'_>` is not implemented for `ConstantValue<'_>`
    |
    = help: the following other types implement trait `ZeroMapKV<'a>`:
              char
              i8
              i16
              i32
              i64
              i128
              usize
              u8
            and 72 others
note: required by a bound in `ZeroMap`
   --> /Users/younies/icu/icu4x/utils/zerovec/src/map/map.rs:64:8
    |
61  | pub struct ZeroMap<'a, K, V>
    |            ------- required by a bound in this struct
...
64  |     V: ZeroMapKV<'a> + ?Sized,
    |        ^^^^^^^^^^^^^ required by this bound in `ZeroMap`

error[E0277]: the trait bound `ConstantValue<'_>: ZeroMapKV<'_>` is not satisfied in `ZeroMap<'_, _, ConstantValue<'_>>`
   --> provider/datagen/src/transform/cldr/units/mod.rs:129:13
    |
129 |         let constants_map = ZeroMap::from_iter(constants_map.into_iter());
    |             ^^^^^^^^^^^^^ within `ZeroMap<'_, _, ConstantValue<'_>>`, the trait `ZeroMapKV<'_>` is not implemented for `ConstantValue<'_>`
    |
    = help: the following other types implement trait `ZeroMapKV<'a>`:
              char
              i8
              i16
              i32
              i64
              i128
              usize
              u8
            and 72 others
note: required because it appears within the type `ZeroMap<'_, _, ConstantValue<'_>>`
   --> /Users/younies/icu/icu4x/utils/zerovec/src/map/map.rs:61:12
    |
61  | pub struct ZeroMap<'a, K, V>
    |            ^^^^^^^
    = note: all local variables must have a statically known size
    = help: unsized locals are gated as an unstable feature

error[E0308]: mismatched types
   --> provider/datagen/src/transform/cldr/units/mod.rs:130:41
    |
130 |         let result = UnitsConstantsV1 { constants_map };
    |                                         ^^^^^^^^^^^^^ expected `ZeroMap<'_, str, ConstantValueULE>`, found `ZeroMap<'_, _, ConstantValue<'_>>`
    |
    = note: expected struct `ZeroMap<'_, str, ConstantValueULE>`
               found struct `ZeroMap<'_, _, ConstantValue<'_>>`

Some errors have detailed explanations: E0277, E0308.
For more information about an error, try `rustc --explain E0277`.
warning: `icu_datagen` (lib) generated 3 warnings
error: could not compile `icu_datagen` (lib) due to 5 previous errors; 3 warnings emitted
warning: build failed, waiting for other jobs to finish...
warning: `icu_datagen` (lib test) generated 3 warnings (3 duplicates)
error: could not compile `icu_datagen` (lib test) due to 5 previous errors; 3 warnings emitted
[cargo-make] ERROR - Error while executing command, exit code: 101
[cargo-make] WARN - Build Failed.
```

Could you help me ?

Reply:
@robertbastian , nice ping

Reply:
Will you be in the office today? We can debug then.

Reply:
I am wfh to day, okay lets do it tomorrow while I am working on something else.

Reply:
Thanks a lot :)